### PR TITLE
Change StackOverflow Job Search

### PR DESCRIPTION
**Pull request by NickyBobby**
Switched over from Feedjira to Nokogiri
Changed all the methods to grab information
Added Nokogiri gem
Added separate location entry for future schema setup
Started rake task to fix locations - WIP

### CODE REVIEW

**Comment by NickyBobby** [comment]
Let me know what you think of these changes. There may be a more elegant solution to using Nokogiri but this was the way I figured it out first. Next thing I need to work on is making a one off rake task to fix locations currently in the database. @rrgayhart @erinnachen @hhoopes @brennanholtzclaw

File: app/services/stack_overflow.rb [hunk 1]
```diff
@@ -9,30 +9,37 @@ def initialize(term="ruby")
...
+    if feed.present?
+      format_feed(feed)
     end
   end
 
-  def format_entries(entries)
-    entries.map do |entry|
-      formatted_entry = self.format_entry(entry)
+  def format_feed(feed)
+    feed.css('item').map do |item|
```
Comment (erinnachen) [inline app/services/stack_overflow.rb]: I'm unsure why you changed the name of this method, especially because this method is pretty consistently called format_entries across all services.

File: app/services/stack_overflow.rb [hunk 2]
```diff
@@ -54,11 +61,13 @@ def is_remote?(title)
...
+    entry.css('category').map do |category|
+      category.text
+    end
   end
 
   def pull_feed
     url = BASE_URI + term
-    Feedjira::Feed.fetch_and_parse url
+    @rss_feed ||= Nokogiri::HTML(open(url))
   end
```
Comment (erinnachen) [inline app/services/stack_overflow.rb]: Why did you choose to memoize this?

File: app/services/stack_overflow.rb [hunk 3]
```diff
@@ -9,30 +9,37 @@ def initialize(term="ruby")
...
-        posted_date: entry.published
+        description: description,
+        remote: self.is_remote?(title),
+        posted_date: entry.css('pubdate').text
       },
       company: {
-        name: self.pull_company_name(entry.title)
+        name: self.pull_company_name(title)
+      },
+      location: {
```
Comment (erinnachen) [inline app/services/stack_overflow.rb]: There doesn't seem to be a need for this currently. You load the location as part of jobs (as in line 32).

File: app/services/stack_overflow.rb (hunk 2 above)
Comment (erinnachen) [inline app/services/stack_overflow.rb]: The feed I think is in XML, do you think it would be easier to use the Nokogiri::XML functionality?

File: app/services/stack_overflow.rb (hunk 1 above)
Comment (NickyBobby) [inline app/services/stack_overflow.rb]: You're right. I don't know why I changed it. I'll change it back.

**Comment by erinnachen** [comment]
@NickyBobby I left questions/comments in the code.

File: app/services/stack_overflow.rb (hunk 2 above)
Comment (NickyBobby) [inline app/services/stack_overflow.rb]: My thought was to memoize it in case we needed to call on it a few different times but we're not doing that, so it's unnecessary.

File: app/services/stack_overflow.rb (hunk 3 above)
Comment (NickyBobby) [inline app/services/stack_overflow.rb]: I added the extra location entry to work with your future implementation of the location table. You're right we don't need it right now but we will soon.

**Comment by NickyBobby** [comment]
Thank you. I'll make those changes right now. @erinnachen

**Comment by NickyBobby** [comment]
I added to the rake task. I think I found a good way to fix the locations of the existing jobs. I even added a few extra conditionals to take care of Singapore and Hong Kong locations. And the way I have it set up right now is if it doesn't match up with any of those things is it will be deleted. What do we think of that? I can change it if we don't want to do that. Let me know @hhoopes @erinnachen @brennanholtzclaw

File: lib/tasks/one_off_tasks/fix_location.rake [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  desc "fix locations currently in the database"
+
+  task all_previous_records: :environment do
+    jobs_with_locations
+  end
+
+  def jobs_with_locations
+    stackoverflow_jobs = Job.where('url LIKE ?', '%stackoverflow%').all
+    stackoverflow_jobs.each do |job|
+      unless job.location =~ /(\w*, \w*)/
```
Comment (erinnachen) [inline lib/tasks/one_off_tasks/fix_location.rake]: @NickyBobby What do you do with locations that don't have a comma in them?

File: lib/tasks/one_off_tasks/fix_location.rake [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        normalize_location(job)
+      end
+    end
+  end
+
+  def normalize_location(job)
+    if job.title =~ /(\w*, \w*)/
+      job.update_attribute(location: job.title[/(\w*, \w*)/])
```
Comment (erinnachen) [inline lib/tasks/one_off_tasks/fix_location.rake]: This seems like a repeat of how the location was originally generated. I'm not quite sure that you will eliminate situations where the title is for example: "Software engineer (JavaScript, Ruby, Go) at Company in (Cool Place, NY)"

**Comment by NickyBobby** [comment]
It's a different type of regex. And I went through all the data and didn't seem to come across any cases that were like (Angular, Ruby) if they were to post the stacks it's generally (Angular,Ruby) with no space. Redoing the regex through the way I did it eliminates most of the "bad location" jobs. The only thing it wouldn't pick up on was Singapore and Hong Kong, so that's why I have a conditional in there. But you are right about it picking up on (Angular, Ruby). I just haven't seen that yet. I was also playing around with the idea of sending all the locations to an API to see if it's a valid location or not. Wasn't sure which one to use tho and most API's do more complex things than just return true or false if it's a valid location. I'm open to suggestions if this isn't a good one to go about doing this. @erinnachen @hhoopes @brennanholtzclaw

**Comment by rrgayhart** [comment]
@NickyBobby - This is great work! Probably the best way to conclusively answer the questions about the matches would be to write up a rake task test: https://robots.thoughtbot.com/test-rake-tasks-like-a-boss

You'd want to run the task on your records locally - write a bunch of tests for records that worked well - then find some records that didn't port correctly and add tests around them and see if you can handle those edge cases without breaking the ones that were working before.

LMK if you have any questions about that!

File: db/schema.rb [hunk 6]
```diff
@@ -15,6 +15,7 @@
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
+  enable_extension "pg_stat_statements"
```
Comment (rrgayhart) [inline db/schema.rb]: What is this feller doing?

File: lib/tasks/one_off_tasks/fix_location.rake [hunk 7]
```diff
@@ -0,0 +1,31 @@
+namespace :fix_location do
+  desc "fix locations currently in the database"
+
+  task all_previous_records: :environment do
+    jobs_with_locations
+  end
+
+  def jobs_with_locations
+    stackoverflow_jobs = Job.where('url LIKE ?', '%stackoverflow%').all
+    stackoverflow_jobs.each do |job|
```
Comment (rrgayhart) [inline lib/tasks/one_off_tasks/fix_location.rake]: Probably want to batch this or use for_each to go through the jobs, so you don't overtax heroku

File: lib/tasks/one_off_tasks/fix_location.rake [hunk 8]
```diff
@@ -0,0 +1,31 @@
+namespace :fix_location do
+  desc "fix locations currently in the database"
+
+  task all_previous_records: :environment do
+    jobs_with_locations
+  end
+
+  def jobs_with_locations
+    stackoverflow_jobs = Job.where('url LIKE ?', '%stackoverflow%').all
```
Comment (rrgayhart) [inline lib/tasks/one_off_tasks/fix_location.rake]: Nice!

**Comment by NickyBobby** [comment]
Yea I'm gonna do that. Im doing some "manual" testing right now through the rails console. And there are some edge cases that are passing through my initial implementation. Do you know of anyway to check if a word is an actual location? Like a database to download and match against or an API? I'm gonna start building some rake task tests in a few minutes. @rrgayhart

**Comment by rrgayhart** [comment]
@NickyBobby - Hrm, good question - you could potentially use the GeoCoder gem to see if it finds the place: http://www.rubygeocoder.com/ --- It's very possible that it won't work for our purposes.

File: db/schema.rb (hunk 6 above)
Comment (NickyBobby) [inline db/schema.rb]: That's a very good question! @rrgayhart

**Comment by NickyBobby** [comment]
I have used the geocoder gem for finding latitude and longitude locations. I'll see if we can use it for our purposes. I just want something to send back a true or false if it exists or not. I'll check'er out. @rrgayhart

**Comment by NickyBobby** [comment]
So I added the geocoder gem, which meant I also needed to add latitude and longitude attributes to the jobs table. I spent a lot of time tonight figuring out the best way to find the locations of each job and I got it down to 295 jobs that don't have proper geocodes, which means I haven't found their locations yet. The rake task is still a work in progress, but one thing of concern is that this might take a while to complete on heroku. It took me over an hour to update our current database in development. Also have to be careful that we don't over the geocode rate limit. There is a way to slow it down tho which might make the transition smooth. Like I said tho, I got the number of missing locations down under 300 so I'll figure out how to fix the rest of the jobs tomorrow. @rrgayhart @erinnachen @hhoopes @brennanholtzclaw

**Comment by NickyBobby** [comment]
Also I had to comment some stuff out of the tests to get them to pass. That's why the coverage went down. I'll work on that tomorrow as well.

**Comment by erinnachen** [comment]
@NickyBobby :Which lookup API are you using? I looked into this last week (probably should have weighed in a little earlier) and I think there were a few that had really high limits. If I remember correctly, a few were sort of gnarlier in that their TOC required some display of their logo/or a map of the geocoded locations. I really liked the [mapquest API](http://www.mapquestapi.com/geocoding/) and they have a batch geocoding request, also there are city, state and country fields as well in the geocoding, so you may not have to store longitude/latitude.

**Comment by erinnachen** [comment]
@NickyBobby here are also some "edge cases" that I found in the location data:
openings in Chicago, Sydney, Amsterdam
React, Redux, Elm, Purescript
Goshen, IN; Tucson, AZ; or Remote
.NET, C#, MVC6, Agile, AngularJS, Azure
GoLang, Go, Ruby
frontend, personalization, profit
Node.js, AngularJS, Mobile

Keep up the good work! I know this is pretty tedious stuff.

**Comment by NickyBobby** [comment]
I'm using the geocode gem which I believe uses the google API to grab the lat and long coords. It's a lot to go through the entire database but probably won't be too bad once we normalize the locations to a separate table. Also I thought it would be cool to have the coordinate information if we ever wanted to display these on a map one day. And yea I found a bunch of edge cases last night as well. I'm gonna have to beef up the regex to weed out those "imposter" locations. @erinnachen

**Comment by erinnachen** [comment]
@NickyBobby @rrgayhart There is a license-restriction located [here](https://developers.google.com/maps/terms#10-license-restrictions) that I was a bit concerned about
10.4.d: **No use of Content without a Google map**. Unless the Maps APIs Documentation expressly permits you to do so, you will not use the Content in a Maps API Implementation without a corresponding Google map. For example, you may display Street View imagery without a corresponding Google map because the Maps APIs Documentation expressly permits this use.
Any thoughts?

File: lib/tasks/one_off_tasks/fix_location.rake [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    elsif job.title.include?('Hong Kong')
+      job.update_attributes(location: "Hong Kong")
+    elsif job.title.include?('remote')
+      job.update_attribute(remote: true)
+      job.update_attribute(location: "remote")
+    end
+  end
+
+  def check_for_location(job)
+    # search other spots for the location
```
Comment (NickyBobby) [inline lib/tasks/one_off_tasks/fix_location.rake]: Let me know if you have any ideas for beefing up this rake task. What I'm doing currently is checking the title for the pattern (word, other_word) then grabbing the contents of that parentheses and searching geocode with that location. If the response comes back empty then I know it's not a valid location and so I'm gonna send it to another method to check somewhere else. I was thinking about possibly checking a second parentheses for that pattern and checking it? Do you have any thoughts on that? @hhoopes @brennanholtzclaw @erinnachen @rrgayhart

Comment (hhoopes) [inline lib/tasks/one_off_tasks/fix_location.rake]: Would it be easier and/or complementary to see if the contents of those parentheses are in our Technology table, and reject those first before running against geocode? That way you can rule out any ( , ) matches that are obviously technology.

Comment (NickyBobby) [inline lib/tasks/one_off_tasks/fix_location.rake]: Yea that's a good idea. I'm gonna work on that first thing in the morning. @hhoopes

**Comment by brennanholtzclaw** [comment]
After looking this over I think it looks good. Let's remember to take out the limitation [here](https://github.com/LookingForMe/lookingfor/pull/74/files#diff-2fa8534f1cbbf7d009bf6f235bfaa324R9) before merging in. Also, can/should we clean up some of the comments in [this file](https://github.com/LookingForMe/lookingfor/pull/74/files#diff-05a4c8e202d7bccb52a2b0b16190ca7fR1) to document just what we need and are using?

**Comment by NickyBobby** [comment]
I updated the rake task and it ACTUALLY works now. There may be more edge cases that I haven't found but I did have to add a simple array of "locations" that were causing some trouble! I'd love to hear if anyone has any feedback about it. Also I added a sleep function after each Geocode call. This should get us around the whole rate limit thang. Next thing I can do is try it on the whole development database and see if there are any other edge cases. @rrgayhart @hhoopes @brennanholtzclaw

**Comment by NickyBobby** [comment]
After a very long time, I am glad to say that this branch is finally ready to be merged. I ran the rake task against the entire development database and haven't seen any funky stuff left for locations. There may still be a few edge cases in there but it looks pretty good. I'm gonna take this off of WIP. I'd like everyone to get a look at it before @hhoopes merges it (we have a "all merge yours if you merge mine" pact going right now). In fact I'm gonna start my very thorough code review of her PR right meow!! @rrgayhart @brennanholtzclaw

**Comment by hhoopes** [comment]
I have a pretty crazy development database so I can run it against that.

**Comment by hhoopes** [comment]
Sorry, forgot to merge this after our conversation. Would like it if I could get some help reviewing erinna's location branch so we can get these closed. @NickyBobby @brennanholtzclaw
